docs/coding-guidelines/vectorization-guidelines.md

```diff
@@ -0,0 +1,1091 @@
+
+```
+------------------------------128-bits---------------------------
+|             64                |               64              |
+-----------------------------------------------------------------
+|      32       |      32       |      32       |      32       |
+----------------------------------------------------------------|
+|  16   |  16   |  16   |  16   |  16   |  16   |  16   |  16   |
+-----------------------------------------------------------------
+| 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 |
+-----------------------------------------------------------------
+```
+
+`Vector256<T>` is twice as big as `Vector128<T>`, so when it is hardware accelerated, and the data is large enough and the benchmarks prove that it offers better performance, you should use it instead of `Vector128<T>`. Namely, `Vector256<T>` on x86/x64 is mostly treated as `2x Vector128<T>` and while there are some operations that can "cross lanes", they can sometimes be more expensive or have other hidden costs.
+
+To check the acceleration, use `Vector128.IsHardwareAccelerated` and `Vector256.IsHardwareAccelerated` properties.
+
+The size of the input also matters. It needs to be at least of the size of a single vector to be able to execute the vectorized code path (there are some advanced tricks that can allow you to operate on smaller inputs, but we won't describe them here). `Vector128<T>.Count` and `Vector256<T>.Count` return the number of elements of the given type T in a single vector.
+Both `Count` and `IsHardwareAccelerated` are turned into constants by the Just-In-Time compiler (i.e. no method call is required to retrieve the information). In the case of pre-compiled code (NativeAOT), this is not true for the `IsHardwareAccelerated` property, as the required information is not available at compile time.
+
+**Note:** When `Vector256` is accelerated then `Vector128` and `Vector64` are also accelerated.
+
+That is why the code is very often structured like this:
+
+```cs
+void CodeStructure(ReadOnlySpan<byte> buffer)
+{
+    if (Vector256.IsHardwareAccelerated && buffer.Length >= Vector256<byte>.Count)
+    {
+        // Vector256 code path
+    }
+    else if (Vector128.IsHardwareAccelerated && buffer.Length >= Vector128<byte>.Count)
+    {
+        // Vector128 code path
+    }
+    else
+    {
+        // non-vectorized && small inputs code path
+    }
+}
+```
+
+To reduce the number of comparisons for small inputs, we can re-arrange it in the following way:
+
+```cs
+void OptimalCodeStructure(ReadOnlySpan<byte> buffer)
+{
+    if (!Vector128.IsHardwareAccelerated || buffer.Length < Vector128<byte>.Count)
+    { 
+        // scalar code path
+    } 
+    else if (!Vector256.IsHardwareAccelerated || buffer.Length < Vector256<byte>.Count)
+    { 
+        // Vector128 code path
+    } 
+    else
+    { 
+        // Vector256 code path
+    }
+}
+```
```

Comment:
Is it worth just giving this as a single example here and pushing devs towards it as the implementation?

The only "downside" is that it may require duplication of logic to handle any "trailing elements" that don't fit into the vector, where-as with the first approach you can have the vector path fall into the `else` path to handle the trailing data to simplify your code.

No preference, there are upsides and downsides to each.